lib/src/source/hosted.dart

```diff
@@ -311,17 +311,27 @@ class BoundHostedSource extends CachedSource {
     versionListing ??= await _cachedVersionListingResponse(ref, maxAge);
     // Otherwise retrieve the info from the host.
     versionListing ??= await _scheduler.schedule(ref);
+
+    final listing = versionListing[id];
     // If we don't have the specific version we return the empty response.
+    //
+    // This should not happen. But in production we want to avoid a crash, since
+    // it is more or less harmless.
+    //
     // TODO(sigurdm): Consider representing the non-existence of the
     // package-version in the return value.
+    assert(listing != null);
     return versionListing[id]?.status ?? PackageStatus();
   }
 
   // The path where the response from the package-listing api is cached.
   String _versionListingCachePath(PackageRef ref) {
     final parsed = source._parseDescription(ref.description);
     final dir = _urlToDirectory(parsed.last);
-    return p.join(systemCacheRoot, dir, '${ref.name}-versions.json');
+    // Use a .versions/ dir because older versions of pub won't choke on that
+    // name when iterating the cache (it is not listed by [listDir]).
+    return p.join(
+        systemCacheRoot, dir, '.versions', '${ref.name}-versions.json');
```

Comment:
nit: Do we need `-versions.json`, I guess it makes sense if we want the ability to add multiple files in the future.
In that case perhaps we should make it `.cache/<pacakge>-versions.json` instead of `.versions/<pacakge>.json`.

Either option works for me, but repeating versions twice seems unnecessary :D